File: macOS/RCForb/Sources/Networking/IpExClient.swift
import Foundation
import Network

/// IpEx NAT traversal client ported from ipex-client.ts
class IpExClient {
    private static let IPEX_HOST = "ipex.remotehams.com"
    private static let IPEX_PORT: UInt16 = 7005

    private var connection: NWConnection?
    private var isConnectedFlag = false
    private var heartbeatTimer: DispatchSourceTimer?
    private var timeoutTimer: DispatchSourceTimer?
    private var lastDataTime = Date()
    private var lineBuffer = ""
    private let queue = DispatchQueue(label: "com.rcforb.ipex")

    var onHolePunchGo: ((String) -> Void)?
    var onHolePunchOk: (() -> Void)?
    var onHolePunchFail: (() -> Void)?
    var onServerNotFound: (() -> Void)?
    var onDisconnected: (() -> Void)?

    var isConnected: Bool { isConnectedFlag }

    func connect() async -> Bool {
        let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(Self.IPEX_HOST),
            port: NWEndpoint.Port(integerLiteral: Self.IPEX_PORT)
        )
        let conn = NWConnection(to: endpoint, using: .tcp)
        self.connection = conn

        return await withCheckedContinuation { continuation in
            conn.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    self?.isConnectedFlag = true
                    self?.lastDataTime = Date()
                    self?.startHeartbeat()
                    self?.startReceiving()
                    continuation.resume(returning: true)
                case .failed, .cancelled:
                    self?.handleDisconnect()
                    continuation.resume(returning: false)
                default: break
                }
            }
            conn.start(queue: queue)

            queue.asyncAfter(deadline: .now() + 10) { [weak self] in
                guard let self, !self.isConnectedFlag else { return }
                conn.cancel()
            }
        }
    }

    func disconnect() {
        isConnectedFlag = false
        stopTimers()
        connection?.cancel()
        connection = nil
        onDisconnected?()
    }

    private func handleDisconnect() {
        guard isConnectedFlag else { return }
        isConnectedFlag = false
        stopTimers()
        connection = nil
        onDisconnected?()
    }

    func holePunchRequest(_ serverEndpoint: String, _ clientPort: Int) {
        send("ClientConnectRequest,\(serverEndpoint),\(clientPort)\n")
    }

    func connectRequestCompleted() { send("ClientConnectRequest,OK\n") }
    func connectRequestFailed() { send("ClientConnectRequest,FAIL\n") }

    private func send(_ data: String) {
        guard let conn = connection, isConnectedFlag else { return }
        conn.send(content: data.data(using: .utf8), completion: .contentProcessed { _ in })
    }

    private func startReceiving() {
        connection?.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, _, error in
            guard let self, let data, error == nil else { return }
            self.processData(String(data: data, encoding: .utf8) ?? "")
            self.startReceiving()
        }
    }

    private func processData(_ data: String) {
        lastDataTime = Date()
        lineBuffer += data
        var lines = lineBuffer.components(separatedBy: "\n")
        lineBuffer = lines.removeLast()
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if !trimmed.isEmpty { processLine(trimmed) }
        }
    }

    private func processLine(_ line: String) {
        if line.hasPrefix("ClientConnectRequest,GO") {
            let parts = line.components(separatedBy: ",")
            let endpoint = parts.count > 2 ? parts[2] : ""
            onHolePunchGo?(endpoint)
        } else if line == "ClientConnectRequest,OK" {
            onHolePunchOk?()
        } else if line == "ClientConnectRequest,FAIL" {
            onHolePunchFail?()
        } else if line.hasPrefix("ServerNotFound") {
            onServerNotFound?()
        }
    }

    private func startHeartbeat() {
        heartbeatTimer = DispatchSource.makeTimerSource(queue: queue)
        heartbeatTimer?.schedule(deadline: .now(), repeating: 4)
        heartbeatTimer?.setEventHandler { [weak self] in
            self?.send("\(Int(Date().timeIntervalSince1970 * 1000))\n")
        }
        heartbeatTimer?.resume()

        timeoutTimer = DispatchSource.makeTimerSource(queue: queue)
        timeoutTimer?.schedule(deadline: .now() + 1, repeating: 1)
        timeoutTimer?.setEventHandler { [weak self] in
            guard let self else { return }
            if Date().timeIntervalSince(self.lastDataTime) > 15 {
                print("[IpEx] Heartbeat timeout")
                self.handleDisconnect()
            }
        }
        timeoutTimer?.resume()
    }

    private func stopTimers() {
        heartbeatTimer?.cancel(); heartbeatTimer = nil
        timeoutTimer?.cancel(); timeoutTimer = nil
    }
}

File: android/app/src/main/java/com/rcforb/android/networking/IpExClient.kt
package com.rcforb.android.networking

import android.util.Log
import kotlinx.coroutines.*
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket

class IpExClient {
    companion object {
        private const val IPEX_HOST = "ipex.remotehams.com"
        private const val IPEX_PORT = 7005
    }

    private var socket: Socket? = null
    private var output: OutputStream? = null
    private var isConnectedFlag = false
    private var lastDataTime = System.currentTimeMillis()
    private var lineBuffer = StringBuilder()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var jobs = mutableListOf<Job>()

    var onHolePunchGo: ((String) -> Unit)? = null
    var onHolePunchOk: (() -> Unit)? = null
    var onHolePunchFail: (() -> Unit)? = null
    var onServerNotFound: (() -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null

    val isConnected: Boolean get() = isConnectedFlag

    suspend fun connect(): Boolean = withContext(Dispatchers.IO) {
        try {
            val sock = Socket(IPEX_HOST, IPEX_PORT)
            sock.soTimeout = 500
            socket = sock
            output = sock.getOutputStream()
            isConnectedFlag = true
            lastDataTime = System.currentTimeMillis()
            startHeartbeat()
            startReceiving()
            true
        } catch (e: Exception) {
            Log.e("IpExClient", "Connect failed", e)
            false
        }
    }

    fun disconnect() {
        isConnectedFlag = false
        jobs.forEach { it.cancel() }
        jobs.clear()
        try { socket?.close() } catch (_: Exception) {}
        socket = null; output = null
        onDisconnected?.invoke()
    }

    private fun handleDisconnect() {
        if (!isConnectedFlag) return
        isConnectedFlag = false
        jobs.forEach { it.cancel() }
        jobs.clear()
        try { socket?.close() } catch (_: Exception) {}
        socket = null; output = null
        onDisconnected?.invoke()
    }

    fun holePunchRequest(serverEndpoint: String, clientPort: Int) {
        send("ClientConnectRequest,$serverEndpoint,$clientPort\n")
    }

    fun connectRequestCompleted() = send("ClientConnectRequest,OK\n")
    fun connectRequestFailed() = send("ClientConnectRequest,FAIL\n")

    private fun send(data: String) {
        if (!isConnectedFlag) return
        try { output?.write(data.toByteArray(Charsets.UTF_8)) } catch (_: Exception) {}
    }

    private fun startReceiving() {
        val job = scope.launch {
            val buf = ByteArray(65536)
            val input: InputStream = socket?.getInputStream() ?: return@launch
            while (isActive && isConnectedFlag) {
                try {
                    val len = input.read(buf)
                    if (len > 0) processData(String(buf, 0, len, Charsets.UTF_8))
                    else if (len == -1) { handleDisconnect(); break }
                } catch (_: java.net.SocketTimeoutException) {
                } catch (_: Exception) { if (isConnectedFlag) handleDisconnect(); break }
            }
        }
        jobs.add(job)
    }

    private fun processData(data: String) {
        lastDataTime = System.currentTimeMillis()
        lineBuffer.append(data)
        val lines = lineBuffer.toString().split("\n").toMutableList()
        lineBuffer = StringBuilder(lines.removeAt(lines.size - 1))
        for (line in lines) {
            val trimmed = line.trim()
            if (trimmed.isNotEmpty()) processLine(trimmed)
        }
    }

    private fun processLine(line: String) {
        when {
            line.startsWith("ClientConnectRequest,GO") -> {
                val parts = line.split(",")
                val endpoint = if (parts.size > 2) parts[2] else ""
                onHolePunchGo?.invoke(endpoint)
            }
            line == "ClientConnectRequest,OK" -> onHolePunchOk?.invoke()
            line == "ClientConnectRequest,FAIL" -> onHolePunchFail?.invoke()
            line.startsWith("ServerNotFound") -> onServerNotFound?.invoke()
        }
    }

    private fun startHeartbeat() {
        val hbJob = scope.launch {
            while (isActive && isConnectedFlag) {
                send("${System.currentTimeMillis()}\n")
                delay(4000)
            }
        }
        jobs.add(hbJob)

        val toJob = scope.launch {
            while (isActive && isConnectedFlag) {
                delay(1000)
                if (System.currentTimeMillis() - lastDataTime > 15000) {
                    Log.w("IpExClient", "Heartbeat timeout")
                    handleDisconnect()
                    break
                }
            }
        }
        jobs.add(toJob)
    }
}
